### PR TITLE
Helpers for null-terminated Utf8

### PR DESCRIPTION
**Pull request by sjindel-google**

### CODE REVIEW

File: lib/src/utf8.dart [hunk 1]
```diff
@@ -0,0 +1,40 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library utf8;
```
Comment (lrhn) [inline lib/src/utf8.dart]: There is no need to have a library name, and you risk conflict if any other library in the same program also has the name "utf8" (which isn't that unlikely).

File: lib/src/utf8.dart [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+import 'dart:convert';
+import 'dart:ffi';
+import 'dart:typed_data';
+
+/// [Utf8] implements conversion between Dart strings and null-termianted
+/// Utf8-encoded "char*" strings in C.
```
Comment (lrhn) [inline lib/src/utf8.dart]: Document how it handles Dart strings containing NUL characters?
(Can be done on the individual functions too).

**Comment by mit-mit** [comment]
LGTM overall, Lasse will do a style review

File: lib/src/utf8.dart [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    for (; string[i] != 0; ++i) {
+      if (i == count) {
+        count *= 2;
+        string = array.asExternalTypedData(count: count);
+      }
+    }
+    return Utf8Decoder().convert(Uint8List.view(string.buffer, 0, i));
+  }
+
+  static Pointer<Utf8> toUtf8(String s) {
```
Comment (lrhn) [inline lib/src/utf8.dart]: Consider giving the argument a longer name than `s`. Maybe just `string` (except that you use that name for something else).

Document the function.

File: lib/src/utf8.dart [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    final List<int> units = Utf8Encoder().convert(s);
```
Comment (lrhn) [inline lib/src/utf8.dart]: I believe `Utf8Encoder.convert` this is now typed to return `Uint8List`. You may want to use that as type.

You can use `const Utf8Encoder()` to avoid allocating a new object each time (if you care).
That's the same object returned by `utf8.encoder` as well, and you could also just write `utf8.encode(s)` for exactly the same result.
Easier to read.

File: lib/src/utf8.dart [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+library utf8;
+
+import 'dart:convert';
+import 'dart:ffi';
+import 'dart:typed_data';
+
+/// [Utf8] implements conversion between Dart strings and null-termianted
+/// Utf8-encoded "char*" strings in C.
+class Utf8 extends Struct<Utf8> {
+  static String fromUtf8(Pointer<Utf8> str) {
```
Comment (lrhn) [inline lib/src/utf8.dart]: Consider naming parameter `string` (avoid abbreviations in general).

Also document of the function, e.g.:

/// Creates a [String] containing the characters UTF-8 encoded in [string].
///
/// The [string] must be a zero-terminated byte sequence of valid
/// UTF-8 encodings of Unicode code points.. 
/// It may also contain UTF-8 encodings of unpaired surrogate code points,
/// which is not otherwise valid UTF-8, but which may be created
/// when encoding a Dart string containing an unpaired surrogate.
///
/// Returns a Dart string containing the decoded code points.

(If that is actually correct, but it suggests what to test).

File: .gitignore [hunk 6]
```diff
@@ -0,0 +1,3 @@
+.packages
+pubspec.lock
+.dart_tool
```
Comment (mkustermann) [inline .gitignore]: sort?

File: lib/ffi.dart [hunk 7]
```diff
@@ -0,0 +1,7 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library ffi;
```
Comment (mkustermann) [inline lib/ffi.dart]: You can consider either prefixing the library name with the package name (I believe they live in a global namespace) or leaving the library declaration away entirely.

File: lib/src/utf8.dart [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    int count = 0x1000;
+    Uint8List string = array.asExternalTypedData(count: count);
+    int i = 0;
+    for (; string[i] != 0; ++i) {
+      if (i == count) {
+        count *= 2;
+        string = array.asExternalTypedData(count: count);
+      }
+    }
+    return Utf8Decoder().convert(Uint8List.view(string.buffer, 0, i));
```
Comment (lrhn) [inline lib/src/utf8.dart]: Can use `const Utf8Decoder()` to avoid allocating new one every time (probably not worth bothering about).
Can also just use `utf8.decode(Uint8List.view(...))`, which is easier to read.

File: lib/src/utf8.dart [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+import 'dart:convert';
+import 'dart:ffi';
+import 'dart:typed_data';
+
+/// [Utf8] implements conversion between Dart strings and null-termianted
+/// Utf8-encoded "char*" strings in C.
+class Utf8 extends Struct<Utf8> {
+  static String fromUtf8(Pointer<Utf8> str) {
+    final Pointer<Uint8> array = str.cast();
```
Comment (mkustermann) [inline lib/src/utf8.dart]: Consider making it explicit that the cast takes a type argument.

File: test/utf8_test.dart [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+
+import 'package:test/test.dart';
+import 'package:ffi/ffi.dart';
+
+main() {
+  test("fromUtf8 . toUtf8 is identity", () {
+    final String start = "Hello World!\n";
+    final String end = Utf8.fromUtf8(Utf8.toUtf8(start));
+    expect(end, equals(start));
+  });
```
Comment (lrhn) [inline test/utf8_test.dart]: This does not actually test that it encodes or decodes correctly (it could be storing UTF-16 in the middle).

Consider testing each function independently, and test some edge cases, like invalid UTF-8 or Dart strings containing unpaired surrogates or NUL characters, or an input of more than 4096 bytes.

File: lib/src/utf8.dart [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+library utf8;
+
+import 'dart:convert';
+import 'dart:ffi';
+import 'dart:typed_data';
+
+/// [Utf8] implements conversion between Dart strings and null-termianted
+/// Utf8-encoded "char*" strings in C.
+class Utf8 extends Struct<Utf8> {
```
Comment (mkustermann) [inline lib/src/utf8.dart]: Given this class has only static members, why should it inherit from `Struct<Utf8>` and why is it not abstract?

It could be just two top-level methods:

```dart
Pointer<Uint8> toNullTerminatedCString(String s) => ...;
String fromNullTerminatedCString(Pointer<Uint8> s) => ...;
```

File: lib/src/utf8.dart [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+  static String fromUtf8(Pointer<Utf8> str) {
+    final Pointer<Uint8> array = str.cast();
+    int count = 0x1000;
```
Comment (mkustermann) [inline lib/src/utf8.dart]: This is really hackisch, please use the normal loads and stores. They should be optimized within a few days.

File: lib/src/utf8.dart [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+    Uint8List string = array.asExternalTypedData(count: count);
+    int i = 0;
+    for (; string[i] != 0; ++i) {
+      if (i == count) {
+        count *= 2;
```
Comment (lrhn) [inline lib/src/utf8.dart]: If I read this correctly, then:
* you increment `i` so that it reaches `count`.
* you read `string[i]` and compare it to 0
* then you check whether `i` is `count`.

That should give an out-of-bounds error on the read before checking. 

You could move the `++i` into the body of the loop, so it's executed before the `i == count` check. Then the loop could just be a `while` loop.

File: lib/src/utf8.dart [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+/// [Utf8] implements conversion between Dart strings and null-termianted
+/// Utf8-encoded "char*" strings in C.
+class Utf8 extends Struct<Utf8> {
+  static String fromUtf8(Pointer<Utf8> str) {
+    final Pointer<Uint8> array = str.cast();
+    int count = 0x1000;
+    Uint8List string = array.asExternalTypedData(count: count);
+    int i = 0;
+    for (; string[i] != 0; ++i) {
+      if (i == count) {
```
Comment (mkustermann) [inline lib/src/utf8.dart]: Did you test this?

We have `string.length == count`. The condition in the for loop will be executed before the body of the for loop. So once `i == count` the `string[i] != 0` should throw index out of bounds exception. Or maybe I'm overlooking something?

File: lib/src/utf8.dart (hunk 1 above)
Comment (sjindel-google) [reply lib/src/utf8.dart]: Done.

File: lib/src/utf8.dart (hunk 2 above)
Comment (sjindel-google) [reply lib/src/utf8.dart]: Done on `toUtf8`. It's not relevant for `fromUtf8`.

File: lib/src/utf8.dart (hunk 13 above)
Comment (sjindel-google) [reply lib/src/utf8.dart]: Done.

File: .gitignore (hunk 6 above)
Comment (sjindel-google) [reply .gitignore]: Done.

File: lib/src/utf8.dart (hunk 1 above)
Comment (sjindel-google) [reply lib/src/utf8.dart]: Done.

File: lib/src/utf8.dart (hunk 3 above)
Comment (sjindel-google) [reply lib/src/utf8.dart]: Done.

File: lib/src/utf8.dart (hunk 4 above)
Comment (sjindel-google) [reply lib/src/utf8.dart]: I don't see the new type:
```dart
abstract class Encoding extends Codec<String, List<int>> {
```

File: lib/src/utf8.dart (hunk 5 above)
Comment (sjindel-google) [reply lib/src/utf8.dart]: Done.

File: lib/ffi.dart (hunk 7 above)
Comment (sjindel-google) [reply lib/ffi.dart]: Removed.

File: lib/src/utf8.dart (hunk 9 above)
Comment (sjindel-google) [reply lib/src/utf8.dart]: Done.

File: lib/src/utf8.dart (hunk 11 above)
Comment (sjindel-google) [reply lib/src/utf8.dart]: The reason to inherit from `Struct` is that we can make a more precise type `Pointer<Utf8>` which is more informative than `Pointer<Uint8>`. It also allows writing native function signatures which have the argument type `Pointer<Utf8>`.

File: lib/src/utf8.dart (hunk 12 above)
Comment (sjindel-google) [reply lib/src/utf8.dart]: I'm not as optimistic. Even if they land today, they still won't be available to Flutter/Dart stable users, who might form a negative opinion of the code. For example https://github.com/dart-lang/sdk/issues/34452#issuecomment-485255442

Either way, it's only an implementation detail of the library.

File: lib/src/utf8.dart (hunk 14 above)
Comment (sjindel-google) [reply lib/src/utf8.dart]: Fixed.

File: lib/src/utf8.dart (hunk 8 above)
Comment (sjindel-google) [reply lib/src/utf8.dart]: Done.

**Comment by sjindel-google** [comment]
Working on better tests.

File: lib/src/utf8.dart (hunk 12 above)
Comment (mit-mit) [reply lib/src/utf8.dart]: Consider opening a tracking issue in this repo to remind us

**Comment by sjindel-google** [comment]
PTAL

File: lib/src/utf8.dart (hunk 12 above)
Comment (mkustermann) [reply lib/src/utf8.dart]: To avoid the weird doubling logic you can use the largest length possible:

```c++
const kMaxSmi64 = (1 << 62) - 1;
const kMaxSmi32 = (1 << 30) - 1;
sizeof<ffi.IntPtr>() == 8 ? kMaxSmi64 : kMaxSmi32;
```

File: lib/src/utf8.dart (hunk 11 above)
Comment (mkustermann) [reply lib/src/utf8.dart]: What does `sizeof<Utf8>()` return? Is it 0? So `Pointer<Utf8>` will be a pointer to a 0-length struct?

That means it's basically just a nominal type representing the address and, depending on how function signatures are generated, might require users to cast this opaque pointer to `Pointer<Int8>` for callbacks or return values.

Maybe it's worth mentioning that in the class description.

File: test/utf8_test.dart [hunk 15]
```diff
@@ -2,13 +2,76 @@
...
 // BSD-style license that can be found in the LICENSE file.
 
+import 'dart:ffi';
+import 'dart:typed_data';
+
 import 'package:test/test.dart';
 import 'package:ffi/ffi.dart';
 
+Pointer<Uint8> _bytesFromList(List<int> ints) {
+  Pointer<Uint8> ptr = Pointer.allocate(count: ints.length);
```
Comment (mkustermann) [inline test/utf8_test.dart]: Prefer final variables if they are not re-assigned

File: lib/src/utf8.dart [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+  /// Returns the length of a null-terminated string -- the number of (one-byte)
+  /// characters before the first null byte.
+  static int strlen(Pointer<Utf8> string) {
+    final Pointer<Uint8> array = string.cast<Uint8>();
+    int count = 0x1000;
+    Uint8List nativeString = array.asExternalTypedData(count: count);
+    int i = 0;
+    while (nativeString[i] != 0) {
+      if (++i == count) {
+        count *= 2;
```
Comment (dcharkes) [inline lib/src/utf8.dart]: Could we use a precise bound here?

File: lib/src/utf8.dart [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+        string.cast<Uint8>().asExternalTypedData(count: length).buffer,
+        0,
+        length));
+  }
+
+  /// Convert a [String] to a Utf8-encoded null-terminated C string.
+  ///
+  /// If 'string' contains NULL bytes, the converted string will be truncated
+  /// prematurely. Unpaired surrogate code points in [string] will be preserved
+  /// in the UTF-8 encoded result. See [Utf8Encoder] for details on encoding
```
Comment (dcharkes) [inline lib/src/utf8.dart]: . at the end of the sentence.

File: test/utf8_test.dart (hunk 15 above)
Comment (lrhn) [reply test/utf8_test.dart]: For the record: That's neither recommended nor discouraged by the style guide, just something some projects may choose to do.

File: test/utf8_test.dart (hunk 10 above)
Comment (sjindel-google) [reply test/utf8_test.dart]: Done.

File: lib/src/utf8.dart (hunk 11 above)
Comment (sjindel-google) [reply lib/src/utf8.dart]: There is no need to cast it in function signatures. I've added a comment.

File: lib/src/utf8.dart (hunk 12 above)
Comment (sjindel-google) [reply lib/src/utf8.dart]: Done.

File: test/utf8_test.dart (hunk 15 above)
Comment (sjindel-google) [reply test/utf8_test.dart]: Done.

File: lib/src/utf8.dart (hunk 16 above)
Comment (sjindel-google) [reply lib/src/utf8.dart]: No, because the length of the string can be only be detected by scanning until the first NULL byte.

File: lib/src/utf8.dart (hunk 17 above)
Comment (sjindel-google) [reply lib/src/utf8.dart]: Done.

File: lib/src/utf8.dart (hunk 16 above)
Comment (dcharkes) [reply lib/src/utf8.dart]: But, why times 2 every time? Why not scan until null and use that as count?

Comment (sjindel-google) [reply lib/src/utf8.dart]: This is scanning until null.

Comment (dcharkes) [reply lib/src/utf8.dart]: Ah, I was confused by `array.asExternalTypedData(count: count)` where that count is potentially 2x the string length. We need an oversized TypedData wrapper in order to not create a new wrapper every character for scanning in the first place. (I hope we can switch to the normal `Pointer<Int8>.load()` soon, that would make it easier to read this code.)
